core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java

```diff
@@ -0,0 +1,180 @@
+        private final ActionListener<IndicesUnassigedShardsResponse> listener;
+
+        AsyncShardFetches(DiscoveryNodes nodes, MetaData metaData, int expectedOps, ActionListener<IndicesUnassigedShardsResponse> listener) {
+            this.nodes = nodes;
+            this.metaData = metaData;
+            this.listener = listener;
+            this.expectedOps = new CountDown(expectedOps);
+            this.fetchResponses = new ConcurrentLinkedQueue<>();
+        }
+
+        void fetch(Collection<ShardId> shardIds) {
+            for (ShardId shardId : shardIds) {
+                InternalAsyncFetch fetch = new InternalAsyncFetch(logger, "monitor_shard_copies", shardId, listShardsInfo);
+                fetch.fetchData(nodes, metaData, Collections.<String>emptySet());
+            }
+        }
+
+        private class InternalAsyncFetch extends AsyncShardFetch<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> {
+
+            InternalAsyncFetch(ESLogger logger, String type, ShardId shardId, TransportNodesListGatewayStartedShards action) {
+                super(logger, type, shardId, action);
+            }
+
+            @Override
+            protected synchronized void processAsyncFetch(ShardId shardId, TransportNodesListGatewayStartedShards.NodeGatewayStartedShards[] responses, FailedNodeException[] failures) {
+                super.processAsyncFetch(shardId, responses, failures);
+                fetchResponses.add(new Response(shardId, responses, failures));
+                if (expectedOps.countDown()) {
+                    finish();
+                }
+            }
+
+            void finish() {
+                ImmutableOpenMap.Builder<String, Map<Integer, java.util.List<ShardStatus>>> shardsResponseBuilder = ImmutableOpenMap.builder();
+                while (!fetchResponses.isEmpty()) {
+                    Response res = fetchResponses.remove();
+                    Map<Integer, java.util.List<ShardStatus>> indexShards = shardsResponseBuilder.get(res.shardId.getIndex());
+                    if (indexShards == null) {
+                        indexShards = new HashMap<>();
+                    }
+                    java.util.List<ShardStatus> shardStatuses = indexShards.get(res.shardId.id());
+                    if (shardStatuses == null) {
+                        shardStatuses = new ArrayList<>();
+                    }
+                    for (TransportNodesListGatewayStartedShards.NodeGatewayStartedShards response : res.responses) {
+                        shardStatuses.add(new ShardStatus(response.getNode().id(), response.version(), response.exception()));
+                    }
+                    indexShards.put(res.shardId.id(), shardStatuses);
+                    shardsResponseBuilder.put(res.shardId.getIndex(), indexShards);
+                }
+                listener.onResponse(new IndicesUnassigedShardsResponse(shardsResponseBuilder.build()));
```

Comment:
now, `IndicesUnassigedShardsResponse` takes in an additional `ImmutableList<FailedNodeException>` aggregating all the failures